### PR TITLE
fix(aws-amplify/auth): debouncing get user session calls

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Short version
This change adds back changes made in https://github.com/aws-amplify/amplify-js/pull/10553, which was reverted in https://github.com/aws-amplify/amplify-js/pull/10630. It was reverted because of DataStore auth integration test failures.
The DataStore issue causes *every* authenticated query/mutation request to make extra Cognito requests.

The additional commit https://github.com/aws-amplify/amplify-js/pull/10654/commits/1fe6949a2a8374add2a0b31ed4a9fc08bfa44293 resolves the DataStore issue.

[Integration test passed here](https://app.circleci.com/pipelines/github/aws-amplify/amplify-js/16701/workflows/fe3d1e48-574e-4621-b972-9ccd36c773dc)

#### Long version

In general this change is to reduce concurrent API requests to CognitoIdentityProvider.GetUser. Compared to previously reverted PR, the additional changes are to resolve the issue that DataStore makes *more* Cognito requests before each query. Here's 2 screenshots of the network tab of an App sign-in a user and `querySubscribe` to a DataStore model. The 2 screenshots shows the difference before and after the change:
<Details>
<Summary>Network tab before the change</Summary>

Screen Shot 2022-11-08 at 12 17 33 PM

</Details>

<Details>
<Summary>Network tab after the change</Summary>

Screen Shot 2022-11-08 at 12 15 27 PM

</Details>

From the screenshots above, you can see not only did the DataStore make more Cognito requests, but also the Authenticator UI is make more Cognito requests. This is likely due to unintended removal localstorage. 

Here's an analogue of the workflow of the DataStore request authentication:
1. For any query/mutation in DataStore module, and the model is marked as cognito user pool authentication, `_headerBasedAuth()` is called: https://github.com/aws-amplify/amplify-js/blob/0fa40e6efb67def43f0eb912b3f2ed34da883f5f/packages/api-graphql/src/GraphQLAPI.ts#L170-L178
2. Fetch the current user pool user in Auth module: https://github.com/aws-amplify/amplify-js/blob/1fe6949a2a8374add2a0b31ed4a9fc08bfa44293/packages/auth/src/Auth.ts#L1573-L1575
    1. First it loads the user session with `user.getSession()`. This function is debounced and only first of concurrent calls are made. The other concurrent invocations will resolve the promise of the 1st invocation: https://github.com/aws-amplify/amplify-js/blob/bbe44eeb29e5461c7a7d85c80f361225ac360851/packages/auth/src/Auth.ts#L1791
    2. With the user session, load the rest of the user attributes with `user.getUserData()`. This function is *not* debounced: https://github.com/aws-amplify/amplify-js/blob/bbe44eeb29e5461c7a7d85c80f361225ac360851/packages/auth/src/Auth.ts#L1639

**What caused more Cognito calls?**
For every DataStore query/mutation invocation, it will reach to the code where we set user session to CognitoUser instance: https://github.com/aws-amplify/amplify-js/blob/bbe44eeb29e5461c7a7d85c80f361225ac360851/packages/auth/src/Auth.ts#L1835 This is necessary because we debounce the concurrent call the fetch the user session. But we still need to set the fetched user session to each of the concurrent CognitoUser instances. If we look deeper to the `user.setSignInUserSession` implementation:https://github.com/aws-amplify/amplify-js/blob/a2380317c3881db596787ffcca3b7afa8211e788/packages/amazon-cognito-identity-js/src/CognitoUser.js#L95-L99. We will find it will remove the user data(including attributes) besides setting the private `signInUserSession` property. As a result, when we call `user.getUserData()`, we always find the user data is disappears from the local storage. Hense, extra Cognito requests need to be made to fetch the user data repeatedly.

**What is the solution?**
The solution is simple, instead of calling `user.setSignInUserSession()`, we can set the private member directly. So the userdata localstorage won't be removed. When `user.getUserData()` is called, it won't make API request, but load from localstorage.

**What is really causing the integration test to fail?**
It's caused by interference between the tests. The each integ test is ended by the [ready event](https://docs.amplify.aws/lib/datastore/datastore-events/q/platform/js/#ready). The event change will trigger new graphql query. Because now each graphql query causes `GetUser` API call and hence refreshes localstorage. By the time previous test ended and 2nd test started, the localstorage is reloaded by the async requests from the 1st test.

### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -1806,47 +1788,66 @@ export class AuthClass {
...
-								await this.cleanUpInvalidSession(user);
-							} catch (cleanUpError) {
-								rej(
-									new Error(
-										`Session is invalid due to: ${err.message} and failed to clean up invalid session: ${cleanUpError.message}`
-									)
-								);
+		const clientMetadata = this._config.clientMetadata;
+		// Debouncing the concurrent userSession calls by caching the promise.
+		// This solution asumes users will always call this function with the same CognitoUser instance.
```
Comment (jimblanc) [inline packages/auth/src/Auth.ts:1798]: I realize original PR had this also, but is this a safe assumption to make? E.g. if `user.getSession` takes longer than normal and in the mean time a different user logs in, would this method return the session from the original user?

Also, small typo `asumes`

**Review by jimblanc — COMMENTED** [review]
Looks good, small question about de-bounce logic. Appreciate the detailed PR write-up!

File: packages/auth/src/Auth.ts (hunk 1 above)
Comment (AllanZhengYP) [reply packages/auth/src/Auth.ts:1798]: Confirmed with @elorzafe this is low risk, but I'd like him to chime-in.

**Review by jimblanc — APPROVED** [review]

File: packages/auth/__tests__/auth-unit-test.ts [hunk 2]
```diff
@@ -1873,6 +1931,47 @@ describe('auth unit test', () => {
...
+					Pool: userPool,
+				});
+				jest.spyOn(user, 'signOut').mockImplementationOnce(() => {
+					userSignoutSpy();
+				});
+				return await auth.userSession(user);
+			});
+			try {
+				await Promise.all(promiseArr);
+				fail('expect promsie to reject');
```
Comment (elorzafe) [inline packages/auth/__tests__/auth-unit-test.ts:1961]: nit
```suggestion
				fail('expect promise to reject');
```

File: packages/auth/__tests__/auth-unit-test.ts [hunk 3]
```diff
@@ -1873,6 +1931,47 @@ describe('auth unit test', () => {
...
+			const credentialsClearSpy = jest.spyOn(Credentials, 'clear');
+			const hubSpy = jest.spyOn(Hub, 'dispatch');
+			let user: CognitoUser | null = null;
+			const getSessionSpy = jest
+				.spyOn(CognitoUser.prototype, 'getSession')
+				.mockImplementationOnce((callback: any) => {
+					callback(new Error('Refresh Token has been revoked'), null);
+				});
+			const userSignoutSpy = jest.fn();
+			// expect.assertions(2);
```
Comment (elorzafe) [inline packages/auth/__tests__/auth-unit-test.ts:1948]: Should we enable this?

File: packages/auth/__tests__/auth-unit-test.ts [hunk 4]
```diff
@@ -1830,6 +1863,31 @@ describe('auth unit test', () => {
 			spyon.mockClear();
 		});
 
+		test('debounce callback error', async () => {
+			const auth = new Auth(authOptions);
```
Comment (elorzafe) [inline packages/auth/__tests__/auth-unit-test.ts:1867]: Is would add here `expect.assertions`

File: packages/auth/__tests__/auth-unit-test.ts [hunk 5]
```diff
@@ -1807,6 +1813,33 @@ describe('auth unit test', () => {
 			spyon.mockClear();
 		});
 
+		test('debouncer happy case', async () => {
```
Comment (elorzafe) [inline packages/auth/__tests__/auth-unit-test.ts:1816]: I would add `expect.assertions`

File: packages/auth/src/Auth.ts (hunk 1 above)
Comment (elorzafe) [reply packages/auth/src/Auth.ts:1798]: is a good concern @jimblanc this class could potenially have a RC on this, do you think we could add in a separate PR a mechanism to use this debouncer for other operations as well?

**Review by elorzafe — CHANGES_REQUESTED** [review]
Just minor comments on tests but overall looks good!

Thanks @AllanZhengYP 🌮

**Review by elorzafe — APPROVED** [review]
Looks good to me

Thanks @AllanZhengYP 🚢

**Review by jimblanc — APPROVED** [review]

**Pull request merged**